{"cells":[{"metadata":{},"cell_type":"markdown","source":"# исключения и файлы\n\n## Исключения\n\nВы уже встречали исключения ранее. Это ситуации, когда что-то идет не так, обычно из-за ошибок в коде или неправильного ввода. Когда происходит исключение, программа немедленно останавливается. \nВ следующем коде происходит исключение ZeroDivisionError при попытке разделить 7 на 0."},{"metadata":{"trusted":true},"cell_type":"code","source":"num1 = 7\nnum2 = 0\nprint(num1/num2)","execution_count":1,"outputs":[{"output_type":"error","ename":"ZeroDivisionError","evalue":"division by zero","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-1-e4e18557d959>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mnum1\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m7\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mnum2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnum1\u001b[0m\u001b[0;34m/\u001b[0m\u001b[0mnum2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m: division by zero"]}]},{"metadata":{},"cell_type":"markdown","source":"Исключения могут происходить по разным причинам.  \nНаиболее частые исключения: \nImportError: импортирование не удалось;  \nIndexError: индекс не входит в диапазон элементов списка;  \nNameError: попытка использовать несуществующую переменную; \nSyntaxError: ошибка разбора кода;  \nTypeError: в функцию передано значение несовместимого типа;  \nValueError: в функцию передано значение совместимого типа, но с некорректным значением."},{"metadata":{},"cell_type":"markdown","source":"В Python есть несколько других стандартных исключений, например, ZeroDivisionError и OSError. Кроме этого, сторонние библиотеки часто имеют свои собственные исключения."},{"metadata":{},"cell_type":"markdown","source":"## Обработка исключений"},{"metadata":{},"cell_type":"markdown","source":"Инструкция try может иметь несколько различных блоков except, чтобы обрабатывать различные исключения. \nБлок except может содержать несколько исключений, которые указываются в круглых скобках; все они будут выполнены программой."},{"metadata":{"trusted":true},"cell_type":"code","source":"try:\n    num1 = 7\n    num2 = 0\n    print (num1 / num2)\n    print(\"Done calculation\")\nexcept ZeroDivisionError:\n    print(\"An error occurred\")\n    print(\"due to zero division\")","execution_count":3,"outputs":[{"output_type":"stream","text":"An error occurred\ndue to zero division\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"Инструкция try может иметь несколько различных блоков except, чтобы обрабатывать различные исключения. \nБлок except может содержать несколько исключений, которые указываются в круглых скобках; все они будут выполнены программой."},{"metadata":{"trusted":true},"cell_type":"code","source":"try:\n    variable = 10\n    print(variable + \"hello\")\n    print(variable / 2)\nexcept ZeroDivisionError:\n    print(\"Divided by zero\")\nexcept (ValueError, TypeError):\n    print(\"Error occurred\")","execution_count":4,"outputs":[{"output_type":"stream","text":"Error occurred\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"Выражение except, не имеющее каких-либо определенных исключений, будет перехватывать все ошибки. Данное выражение следует использовать с осторожностью, так как оно может перехватывать какие-либо неожиданные ошибки и скрыть допущенные ошибки в коде."},{"metadata":{"trusted":true},"cell_type":"code","source":"try:\n    word = \"spam\"\n    print(word / 0)\nexcept:\n    print(\"An error occurred\")","execution_count":5,"outputs":[{"output_type":"stream","text":"An error occurred\n","name":"stdout"}]},{"metadata":{"trusted":true},"cell_type":"code","source":"pin = input()\n\"\"\"\nНам нужно создать программу, которая позволяет создавать свои собственные PIN-коды для банковских карточек. Каждый PIN-код состоит из цифр. Дополните программу, чтобы в случае, когда пользователь вводит буквы, программа останавливалась и выводила \"Please enter a number\", а когда пользователь вводит только цифры, программа выводила \nPIN code is created\". \n \nПример ввода \n44a5 \n \nПример вывода \nPlease enter a number\n\n\"\"\" \ntry:\n    variable = int(pin)\n    print(\"PIN code is created\") \nexcept ValueError: \n    print(\"Please enter a number\")","execution_count":6,"outputs":[{"output_type":"stream","name":"stdout","text":"asda216\nPlease enter a number\n"}]},{"metadata":{},"cell_type":"markdown","source":"## finally\n\nКогда нужно, чтобы некоторый фрагмент кода выполнялся, независимо от того, возникают ошибки или нет, используйте инструкцию finally. Инструкция finally располагается в нижней части инструкции try/except. Инструкция finally выполняется всегда после выполнения блока try и, возможно, после блока except."},{"metadata":{"trusted":true},"cell_type":"code","source":"try:\n    print(\"Hello\")\n    print(1 / 0)\nexcept ZeroDivisionError:\n    print(\"Divided by zero\")\nfinally:\n    print(\"This code will run no matter what\")","execution_count":7,"outputs":[{"output_type":"stream","text":"Hello\nDivided by zero\nThis code will run no matter what\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"Код в инструкции finally будет выполняться, даже если произойдет неперехваченное исключение в одном из предыдущих блоков кода."},{"metadata":{"trusted":true},"cell_type":"code","source":"try:\n    print(1)\n    print(10 / 0)\nexcept ZeroDivisionError:\n    print(unknown_var)\nfinally:\n    print(\"This is executed last\")","execution_count":8,"outputs":[{"output_type":"stream","text":"1\nThis is executed last\n","name":"stdout"},{"output_type":"error","ename":"NameError","evalue":"name 'unknown_var' is not defined","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-8-f051bad4ea7c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;32mexcept\u001b[0m \u001b[0mZeroDivisionError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mZeroDivisionError\u001b[0m: division by zero","\nDuring handling of the above exception, another exception occurred:\n","\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-8-f051bad4ea7c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m10\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mexcept\u001b[0m \u001b[0mZeroDivisionError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0munknown_var\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"This is executed last\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mNameError\u001b[0m: name 'unknown_var' is not defined"]}]},{"metadata":{"trusted":true},"cell_type":"code","source":"coffee = [\"Café Latte\", \"Caffe Americano\", \"Espresso\", \"Cappuccino\", \"Macchiato\"] \n \nchoice = int(input()) \n\"\"\"\nУ каждого вида кофе свой номер, начиная с 0. Напишите программу, \nкоторая получит число от пользователя в качестве ввода и предоставит соответствующий вид кофе. \nЕсли пользователь вводит число вне обозначенного диапазона, программа должна вывести \"Invalid number\". \nНезависимо от результата программа должна вывести \"Have a good day\" в конце. \nПример ввода \n7 \n \nПример вывода \nInvalid number \nHave a good day\n\"\"\" \ntry: \n # место для вашего кода \n    print(coffee[choice]) \nexcept: \n    print(\"Invalid number\") \n  \nfinally: \n #и наконец здесь \n    print(\"Have a good day\")","execution_count":10,"outputs":[{"output_type":"stream","name":"stdout","text":"3\nCappuccino\nHave a good day\n"}]},{"metadata":{},"cell_type":"markdown","source":"## Вызов исключений\n\nВы можете вызывать исключения с помощью инструкции raise.\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"print(1)\nraise ValueError\nprint(2)","execution_count":11,"outputs":[{"output_type":"stream","text":"1\n","name":"stdout"},{"output_type":"error","ename":"ValueError","evalue":"","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)","\u001b[0;32m<ipython-input-11-990863ff3a0f>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mValueError\u001b[0m: "]}]},{"metadata":{},"cell_type":"markdown","source":"Нужно указать тип исключения, которое будет вызвано.\n\nИсключение можно вызвать с указанием аргумента, который представляет собой ее описание."},{"metadata":{"trusted":true},"cell_type":"code","source":"name = \"123\"\nraise NameError(\"Invalid name!\")","execution_count":12,"outputs":[{"output_type":"error","ename":"NameError","evalue":"Invalid name!","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)","\u001b[0;32m<ipython-input-12-d5ceb78ca603>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"123\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32mraise\u001b[0m \u001b[0mNameError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Invalid name!\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mNameError\u001b[0m: Invalid name!"]}]},{"metadata":{},"cell_type":"markdown","source":"В блоках except инструкция raise может использоваться без аргументов, тогда будет повторно вызвано предыдущее исключение."},{"metadata":{"trusted":true},"cell_type":"code","source":"try:\n    num = 5 / 0\nexcept:\n    print(\"An error occurred\")\n    raise","execution_count":13,"outputs":[{"output_type":"stream","text":"An error occurred\n","name":"stdout"},{"output_type":"error","ename":"ZeroDivisionError","evalue":"division by zero","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mZeroDivisionError\u001b[0m                         Traceback (most recent call last)","\u001b[0;32m<ipython-input-13-a238c1d4ad1a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m     \u001b[0mnum\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m5\u001b[0m \u001b[0;34m/\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0;32mexcept\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"An error occurred\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m     \u001b[0;32mraise\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mZeroDivisionError\u001b[0m: division by zero"]}]},{"metadata":{},"cell_type":"markdown","source":"## Утверждения\n\nУтверждение - это проверка правильности кода; ее можно включить или выключить по завершению тестирования программы. \n\nВыражение проверяется, и если оно ложное, вызывается исключение. \nУтверждения создаются с помощью инструкции assert."},{"metadata":{"trusted":true},"cell_type":"code","source":"print(1)\nassert 2 + 2 == 4\nprint(2)\nassert 1 + 1 == 3\nprint(3)","execution_count":14,"outputs":[{"output_type":"stream","text":"1\n2\n","name":"stdout"},{"output_type":"error","ename":"AssertionError","evalue":"","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)","\u001b[0;32m<ipython-input-14-98aeb3b8a49e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32massert\u001b[0m \u001b[0;36m2\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m2\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m4\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m \u001b[0;32massert\u001b[0m \u001b[0;36m1\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;36m3\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n","\u001b[0;31mAssertionError\u001b[0m: "]}]},{"metadata":{},"cell_type":"markdown","source":"Программисты часто размещают утверждения в заголовке функции, чтобы проверить правильность ввода, а также после вызова функции для проверки правильности вывода.\n\nУтверждениям можно давать второй аргумент, который передается в блок AssertionError, выполняемый в случае ложности утверждения.\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"temp = -10\nassert (temp >= 0), \"Colder than absolute zero!\"","execution_count":15,"outputs":[{"output_type":"error","ename":"AssertionError","evalue":"Colder than absolute zero!","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mAssertionError\u001b[0m                            Traceback (most recent call last)","\u001b[0;32m<ipython-input-15-b5d283b4e7fd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mtemp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m-\u001b[0m\u001b[0;36m10\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0;32massert\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mtemp\u001b[0m \u001b[0;34m>=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Colder than absolute zero!\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m","\u001b[0;31mAssertionError\u001b[0m: Colder than absolute zero!"]}]},{"metadata":{},"cell_type":"markdown","source":"Исключения AssertionError могут быть перехвачены и обработаны, как и любое другое исключение, с помощью инструкции try/except. Если же AssertionError не обрабатывается, этот класс исключений приводит к остановке программы.\n"},{"metadata":{},"cell_type":"markdown","source":"## Открытие файлов\nС помощью Python можно читать и редактировать содержимое файлов. \nЛегче всего работать с текстовыми файлами. Перед редактированием файл нужно открыть, что можно сделать с помощью функции open.\n\n### myfile = open(\"filename.txt\")\n\nАргумент функции open - путь к файлу. Если файл находится в текущей рабочей директории программы, можно задать только его имя.\n\nВы можете указать режим, в котором нужно открыть файл, добавив второй аргумент в заголовок функции open. \nЕсли указать \"r\", файл будет открыт в режиме чтения; этот режим используется по умолчанию;  \n\"w\" - файл будет открыт в режиме записи; \n\"а\" - файл будет открыт в режиме добавления: новый контент будет добавляться в конец файла; \n \n\"b\" - файл будет открыт в двоичном режиме, который используется для нетекстовых файлов (таких как изображения и звуковые файлы).\n\n### write mode \nopen(\"filename.txt\", \"w\") \n \n### read mode \nopen(\"filename.txt\", \"r\") \nopen(\"filename.txt\") \n \n### binary write mode \nopen(\"filename.txt\", \"wb\")\n\nВы можете использовать знак + в каждом из режимов выше, чтобы предоставить им дополнительный доступ к файлам. Например, r+ открывает файл для чтения и записи.\n\nПосле того, как вы открыли файл и поработали с ним, его нужно закрыть. \nЭто делается с помощью метода close, указанного в качестве объекта файла.\n\n\nfile = open(\"filename.txt\", \"w\") \n##### do stuff to the file \nfile.close()\n\n\nМы научимся читать/записывать контент файлов немного позже."},{"metadata":{},"cell_type":"markdown","source":"## Чтение данных из файлов\n\nСодержимое файла, открытого в режиме чтения, можно получить с помощью метода read.\n\nfile = open(\"filename.txt\", \"r\") \ncont = file.read() \nprint(cont) \nfile.close()\n\nПрограмма выведет все содержимое файла \"filename.txt\".\n\nЧтобы прочесть только часть данных из файла, в качестве аргумента функции read укажите определенное число. Это количество байтов, которые будут прочитаны.  \n\nПри добавлении большего количества команд чтения с того же файлового объекта, содержимое файла будет прочитано дальше байт за байтом. Если не указать ни одного аргумента, read возвращает всю оставшуюся часть файла.\n\nfile = open(\"filename.txt\", \"r\") \nprint(file.read(16)) \nprint(file.read(4)) \nprint(file.read(4)) \nprint(file.read()) \nfile.close()\n\nКак и без аргументов, отрицательные значения возвращают все содержимое.\n\nПосле того как все содержимое файла было прочитано, дальнейшие вызовы метода чтения данных из файла будут возвращать пустую строку, потому что в данном случае, чтение данных будет производиться с конца файла.\n\nfile = open(\"filename.txt\", \"r\") \nfile.read() \nprint(\"Re-reading\") \nprint(file.read()) \nprint(\"Finished\") \nfile.close()\n\nРезультат :\n\nRe-reading \n \nFinished \n\n\nЧтобы получить каждую строку содержимого файла, используйте метод readlines, который возвращает список, где каждый элемент является строкой файла.\n\n###### Например :\nfile = open(\"filename.txt\", \"r\") \nprint(file.readlines()) \nfile.close()\n\n###### Результат:\n\n['Line 1 text \\n', 'Line 2 text \\n', 'Line 3 text'] \n\n\nТакже можно использовать цикл for для перебора строк файла:\n\n###### Например:\nfile = open(\"filename.txt\", \"r\") \n \nfor line in file: \n    print(line) \n \nfile.close()\n\n###### Результат:\n\nLine 1 text \n \nLine 2 text \n \nLine 3 text \n\n\nВ выведенном результате можно увидеть, что значения разделены пустыми строками, так как функция print автоматически добавляет новую строку после очередного вывода.\n\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"file = open(\"newfile.txt\", \"w\")\nfile.write(\"0-This has been written to a file\\n\")\nfile.write(\"1-This has been written to a file\\n\")\nfile.write(\"2-This has been written to a file\\n\")\nfile.write(\"3-This has been written to a file\\n\")\nfile.write(\"4-This has been written to a file\\n\")\nfile.write(\"5-This has been written to a file\")\nfile.close()\n\nfile = open(\"newfile.txt\",\"r\") \nn = int(input()) \n\"\"\"\nТом подтягивался каждый день и записывал свои результаты. \nОн записывал результат каждого дня в новой строке, \nпоэтому каждая строка представляет собой отдельный день, когда он делал подтягивания. \nНапишите программу, которая принимает число n в качестве входных данных \nи выводит результат n-ного дня (начиная с 0). \n \nПример ввода \n4 \n \nПример вывода \nDay 4, 9 pull ups\n\nИспользуйте функцию readlines(), которая возвращает список, \nсодержащий каждую строку в файле как элемент списка.\n\"\"\" \n#ваш код\ni = 0\nfor line in file:\n    if i == n:\n        print(line)\n    i +=1 \nfile.close()","execution_count":20,"outputs":[{"output_type":"stream","name":"stdout","text":"2\n2-This has been written to a file\n\n"}]},{"metadata":{},"cell_type":"markdown","source":"## Запись в файл\n\nЗапись в файл осуществляется методом write. В файл будет записана строка кода.\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"file = open(\"newfile.txt\", \"w\")\nfile.write(\"This has been written to a file\")\nfile.close()\n\nfile = open(\"newfile.txt\", \"r\")\nprint(file.read())\nfile.close()","execution_count":21,"outputs":[{"output_type":"stream","text":"This has been written to a file\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"Когда файл открывается в режиме записи, все существующее содержимое файла удаляется."},{"metadata":{"trusted":true},"cell_type":"code","source":"file = open(\"newfile.txt\", \"w\")\nfile.write(\"Some new text\")\nfile.close()\n\nfile = open(\"newfile.txt\", \"r\")\nprint(\"Reading new contents\")\nprint(file.read())\nprint(\"Finished\")\nfile.close()","execution_count":22,"outputs":[{"output_type":"stream","text":"Reading new contents\nSome new text\nFinished\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"Метод write в случае успеха возвращает количество байт, записанных в файле.\n"},{"metadata":{"trusted":true},"cell_type":"code","source":"msg = \"Hello world!\"\nfile = open(\"newfile.txt\", \"w\")\namount_written = file.write(msg)\nprint(amount_written)\nfile.close()","execution_count":23,"outputs":[{"output_type":"stream","text":"12\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"Чтобы написать что-то другое, отличное от строки, оно сначала должно быть преобразовано в строку."},{"metadata":{"trusted":true},"cell_type":"code","source":"names = [\"John\", \"Oscar\", \"Jacob\"] \n\"\"\"\nВыполните программу для создания файла, где вы записываете имена из списка, \nкаждое в отдельной строке, и выводите их по отдельности. \n \nВывод \nJohn \nOscar \nJacob\n\"\"\" \nfile = open(\"names.txt\", \"w\") \n#write down the names into the file \nfor i in names:\n    file.write( i + \"\\n\") \nfile.close() \n \nfile= open(\"names.txt\", \"r\") \n#output the content of file in console \nprint(file.read()) \n \nfile.close()","execution_count":24,"outputs":[{"output_type":"stream","text":"John\nOscar\nJacob\n\n","name":"stdout"}]},{"metadata":{},"cell_type":"markdown","source":"## Работаем с файлами\n\nС целью рационального использования ресурсов обычно рекомендуется всегда закрывать файлы после работы с ними. Для этого можно использовать try и finally.\n\ntry: \n   f = open(\"filename.txt\") \n   print(f.read()) \nfinally: \n   f.close()\n   \nТак вы уверенны, что файл всегда будет закрыт, даже если произойдет ошибка.\n  \nЕще один способ сделать это - с помощью инструкции with. Создается временная переменная (часто называемая f), размещаемая в ветви инструкции with.\n\nwith open(\"filename.txt\") as f: \n   print(f.read())\n   \nФайл автоматически закрывается после выполнения инструкции with, даже если возникают исключения.\n"},{"metadata":{},"cell_type":"markdown","source":"### Заголовки книг"},{"metadata":{"trusted":true},"cell_type":"code","source":"file = open(\"newfile.txt\", \"w\")\nfile.write(\"Harry Potter\\n\")\nfile.write(\"Kemal Sunal\\n\")\nfile.write(\"Serdar Ortac\\n\")\nfile.write(\"Killa Hakan\\n\")\nfile.close()\nfile = open(\"newfile.txt\", \"r\") \n\"\"\"\nВам необходимо создать специальную программу для категоризации, которая будет присваивать каждой книге специальный код в зависимости от ее заголовка: \nКод представляет собой первую букву названия книги и количество символов в заголовке. \nНапример, код для книги «Harry Potter» будет выглядеть следующим образом: H12, так как заголовок состоит из 12 символов (включая пробелы) \n \nВам дан файл формата books.txt, в отдельных строках указаны заголовки книг. \nВам нужно прочитать заголовки один за одним и в отдельных строках выдать код каждой книги. \n  \nНапример, если файл формата books.txt содержит такие заголовки: \nSome book \nAnother book \n \nВаша программа должна выдать результат: \nS9 \nA12\n\nВспомните метод readlines(), который возвращает список, содержащий строки файла. \nТакже помните, что все строки, за исключением последней содержат \\n в конце и его не нужно учитывать при подсчете символов.\n\"\"\" \n#введите код сюда \nfor line in file:\n    print(line[0] + str(len(line.replace(\"\\n\",\"\"))) ) \nfile.close()","execution_count":25,"outputs":[{"output_type":"stream","text":"H12\nK11\nS12\nK11\n","name":"stdout"}]}],"metadata":{"kernelspec":{"name":"python3","display_name":"Python 3","language":"python"},"language_info":{"name":"python","version":"3.6.13","mimetype":"text/x-python","codemirror_mode":{"name":"ipython","version":3},"pygments_lexer":"ipython3","nbconvert_exporter":"python","file_extension":".py"}},"nbformat":4,"nbformat_minor":5}